Smarter Loot Roll Logic for Playerbots

Wishmaster117 opened the pull request:
### Summary

This PR improves the loot roll decision-making for Playerbots.

This update introduces spec-aware, stat-aware, and etiquette-friendly rules to better mimic real player behavior.

**Caution, you have to set this config to 2 : AiPlayerbot.LootRollLevel = 2**

### Key Changes
---
**Smart NEED by spec & stats**

- New SmartNeedBySpec system checks whether an item’s stats align with the bot’s main specialization.
- Physical classes (warriors, rogues, hunters, DKs, ferals, ret, enh, prot) never NEED caster gear (SP/INT/SPI/MP5).
- Casters/healers never NEED pure melee DPS gear (STR/AP/ARP/EXP without SP/INT).
- Edge cases (jewelry, trinkets, off-armor) are filtered accordingly.
---
**Cross-armor upgrade logic**

- NEED on off-armor pieces (cloth for paladin, leather for hunter, etc.) only allowed if the new score is significantly higher (CrossArmorExtraMargin).
---
**BoP / BoE / BoU etiquette**

- BoP: allow NEED if the item fills an empty slot and matches spec.
- BoE / BoU: default to GREED, unless explicitly enabled in config.
---
**Token handling (T7–T10)**

- Bots NEED tokens only if they represent a likely upgrade (ilvl + configurable margin).
- Otherwise, they GREED. Unknown-slot tokens default to GREED.
---
**Unique-equip items**

- Bots never NEED duplicates if already equipped or in bags.
---
** Locked chests**
- Rogue lockpicking -> if they have the hability rogues always need in locked chest.
 ---
**Weapon handling**
- So that NEED rolls are more consistent.
---
**System for tank-only loot**
- Prevent DPS from rolling NEED on items with defense, parry, etc.
---
**Disenchant button (Need Before Greed)**

- If enabled, bots choose disenchant on loot if the bot is enchanting
---
**Professions recipes**
- If bot know the profession, have required level and don't know or have the recipe it don NEED otherwise GREED (You can enable or disable rank check and NEED in recipes in the config file)
---
**Special item verification via stat-pattern hook**
This introduces a small, modular hook in the loot roll logic that detects special stat patterns and adjusts NEED/GREED accordingly without disrupting existing rules.

- Pattern registry (extensible): a lightweight matcher/decider path that runs at the start of IsPrimaryForSpec.
- First implemented pattern: INT + Attack Power (incl. Ranged AP).
- Priority classes for INT+AP: Retribution Paladin, Protection Paladin, Hunter, and Enhancement Shaman are treated as primary users and can NEED normally.
- Caster fallback: casters are allowed to NEED only if:
there’s no priority bot in the group, or
there are priority bots present but none is likely to need (checked via each bot’s ItemUsage being EQUIP/REPLACE).

- Humans ignored: human players are not considered in the priority/likelihood checks (they can decide on their own).
- No behavior regressions: BoE/BoU etiquette, BoP slot-fill, recipes, uniques/duplicates, lockboxes, tokens, DE flow, and loot strategy filters remain unchanged.
---
**Roll choice announcement**

- Optional whisper to master (AnnounceToMaster) to see each bot’s roll decision.

---

### New Config Options (with recommended defaults)
```
# NEED only if the item matches bot's main spec (stats & armor/weapon type).
# Recommended: 1
AiPlayerbot.Roll.SmartNeedBySpec = 1

# By default, GREED on Bind-on-Equip unless enabled.
# Recommended: 0
AiPlayerbot.Roll.AllowBoENeedIfUpgrade = 0

# By default, GREED on Bind-on-Use unless enabled.
# Recommended: 0
AiPlayerbot.Roll.AllowBoUNeedIfUpgrade = 0

# Cross-armor NEED allowed only if newScore >= oldScore * margin.
# Recommended: 1.20 (conservative). Use 9.99 to forbid off-armor NEED.
AiPlayerbot.Roll.CrossArmorExtraMargin = 1.20

# If item is DE-only and NBG is active, bots use DE button.
# Recommended: 1
AiPlayerbot.Roll.UseDEButton = 1

# ilvl margin to treat a token as a likely upgrade.
# Recommended: 0.1 (prevents NEED on same-tier duplicates).
AiPlayerbot.Roll.TokenILevelMargin = 0.1

# Whisper each bot's roll choice to its master.
# Recommended: 0
AiPlayerbot.Roll.AnnounceToMaster = 0

```

### Why This Is Better

- Bots now respect class/spec gear logic, reducing frustration when farming with AI.
- More realistic raid etiquette: avoids bots ninja-needing BoEs, caster trinkets, or off-armor.
- Configurable thresholds let admins tune behavior between strict raid-style or lenient casual.
- Improves fairness, immersion, and reduces overhead handling loot drama with bots.

It may respond at this issue : https://github.com/liyunfan1223/mod-playerbots/issues/1592

Wishmaster117 commented:
This is the first realese, it works well, but I need it to be tested, and I will make adjustments if necessary.

Celandriel commented:
Im concerned about config bloat when 1 parameter becomes 6. The config is already confusing enough as it is. 

Im not entirely sure about announcerollchoice implementation. it seems redundant with the 'roll [item]' command that currently exists. In a raid youll just get spammed with loot options, while if the ML provides the item in chat with the appropriate command its more normal.

Celandriel commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -12,6 +12,304 @@
+        return false;
+    };
+
+    // Flags class/spec
```
This section here looks overly complicated for what you are trying to do and ignores existing commands that can do the job more efficiently. 

I suggest branching out the specific function to identify the bot main spec. In that function use existing functions to ID the spec. with early exists the way you have them

```
const uint8 cls = bot->getClass();
if (cls == CLASS_MAGE || cls == CLASS_WARLOCK || cls == CLASS_PRIEST);
 return isPureCasterClass;
if (cls == CLASS_HUNTER || cls == CLASS_ROGUE); 
 return isPhysicalSpec;

if (botAI->IsCaster()) 
  
if (botAI->IsTank())

 if (botAI->IsHealer())  


//fallback case
```

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -618,7 +639,8 @@ bool MasterLootRollAction::Execute(Event event)
     else
     {
-        vote = CalculateRollVote(proto, randomPropertyId ? (int32)randomPropertyId : (randomSuffix ? -(int32)randomSuffix : 0));
+        //vote = CalculateRollVote(proto, randomPropertyId ? (int32)randomPropertyId : (randomSuffix ? -(int32)randomSuffix : 0));
```
Consider deleting commented code

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -461,7 +482,7 @@ RollVote LootRollAction::CalculateRollVote(ItemTemplate const* proto, int32 rand
         if (isPhysicalSpec && looksCaster)
             vote = GREED; // force GREED for rogue/hunter/warrior/DK/retribution/enhancement/feral/protection, etc.
-    }
+    }*/
```
Consider deleting commented code

Wishmaster117 commented:
Ok ,
I’ve reworked the code a bit to make it more readable and added additional checks, especially for weapons. With the code structured this way, once merged, it will be easy to add other checks and conditions that may still be missing.

I’ve also included locked chests for rogue lockpicking, added weapon handling so that NEED rolls are more consistent, and implemented a system for tank-only loot to prevent DPS from rolling NEED on items with defense, parry, etc.

Once the PR is merged, a few adjustments may be needed for very specific cases, but they will be easy to implement.

Wishmaster117 commented:
> Tested on a level 49 mage, everything seems to be well. they greeded when appropriate, didn't see any needs or anything yet. Still doing further testing with it. So far much better then just the whole "passing" on loot and I get everything. The fun of dungeons / raids was the gamble of winning an item you want.

My idea is for bots to act like real players on loot. If this PR is merged, I will improve the mechanics gradually following the various reports that I will receive.

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -47,6 +142,117 @@ static bool HasAnyStat(ItemTemplate const* proto,
+
+// Weapon/shield/relic whitelist per class.
+// Returns false when the item is a WEAPON / SHIELD / RELIC the class should NOT use.
+static bool IsWeaponOrShieldOrRelicAllowedForClass(SpecTraits const& T, ItemTemplate const* proto)
```
Why one method argument is upper letter and second start with lower letter? 
And T is wrong name for variable. Tells nothing what is inside.

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -148,66 +334,101 @@ static bool IsPrimaryForSpec(Player* bot, ItemTemplate const* proto)
     // Class/spec specific adjustments (readable)
     // DK Unholy (DPS): allows STR/HIT/HASTE/CRIT/ARP; rejects all caster items
-    if (isDKUH)
+    if (/* DK Unholy */ (T.cls == CLASS_DEATH_KNIGHT && (T.spec == "unholy" || T.spec == "uh")))
```
Comment above that is condition for DK Unholy is enough

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -547,8 +758,11 @@ RollVote LootRollAction::CalculateRollVote(ItemTemplate const* proto, int32 rand
-    return StoreLootAction::IsLootAllowed(proto->ItemId, GET_PLAYERBOT_AI(bot)) ? vote : PASS;
+    // return StoreLootAction::IsLootAllowed(proto->ItemId, GET_PLAYERBOT_AI(bot)) ? vote : PASS;
+    // Final filter: loot strategy (protect a potential nil AI earlier on login
+    if (PlayerbotAI* ai = GET_PLAYERBOT_AI(bot))
```
Action have directly botAI and i dont see possibility to execute action when botAI == nullptr

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -23,13 +23,108 @@
+            return true;
+    }
+    // Fallback on localized name check
+    static const std::vector<std::string> lockboxKeywords = {
```
I think playerbots support only english dbc

kadeshar commented on conf/playerbots.conf.dist:
```diff
@@ -288,8 +288,8 @@ AiPlayerbot.SayWhenCollectingItems = 1
 # Bots' loot roll level (0 = pass, 1 = greed, 2 = need)
-# Default: 1 (greed), recommanded 2 with new loot AI 
-AiPlayerbot.LootRollLevel = 1
+# Default: 2 (need), recommanded 2 with new loot AI
```
Typo: recommended
Dont understand that description. Why is 2 recommended not 1? Better is highlight if 1 is realized by other methods than 2.

kadeshar replied:
Ok then put that explanation to config file

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -103,53 +939,226 @@ bool LootRollAction::Execute(Event event)
 
-    return StoreLootAction::IsLootAllowed(proto->ItemId, GET_PLAYERBOT_AI(bot)) ? needVote : PASS;
+    // Message to Master
+    botAI->TellMaster(ss.str());
```
No tranlsation and possibility to turn off text. Not all likes spam on chat.

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -12,6 +12,818 @@
+    // Required rank check (can be disabled by config)
+    if (!sPlayerbotAIConfig->recipesIgnoreSkillRank)
+    {
+        if (reqRank && bot->GetSkillValue(reqSkill) < reqRank)
```
Add that to previous if

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -12,6 +12,818 @@
+    // DK Unholy (DPS): allows STR/HIT/HASTE/CRIT/ARP; rejects all caster items
+    if (traits.cls == CLASS_DEATH_KNIGHT && (traits.spec == "unholy" || traits.spec == "uh"))
+    {
+        if (looksCaster) return false;
```
Move that to previous if

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -12,6 +12,818 @@
+    // DK Blood/Frost tanks: DEF/AVOID/STA/STR are useful; reject caster items
+    if (traits.isDKTank)
+    {
+        if (looksCaster) return false;
```
Move that to previous if

Wishmaster117 replied:
Done

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -12,6 +12,818 @@
+    // Rogue (all specs): same strict physical filter (no caster items)
+    if (traits.isRogue)
+    {
+        if (looksCaster) return false;
```
Merge all if's if possible

Wishmaster117 replied:
Done

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -103,53 +939,226 @@ bool LootRollAction::Execute(Event event)
+    // DEBUG: dump for recipes
+    if (IsRecipeItem(proto))
+        DebugRecipeRoll(bot, proto, usage, recipeChecked, recipeUseful, recipeKnown,
+                        reqSkillDbg, reqRankDbg, botRankDbg, /*before*/ (recipeNeed?NEED:PASS), /*after*/ finalVote);
```
Multiline statement in if without {}

kadeshar commented:
It is not marked as draft and every review you adding something new. Last time i reviewing such changes. PR means you ended work.

noisiver commented:
Marking as draft until it is ready to be tested and reviewed.

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -943,11 +943,11 @@ bool LootRollAction::Execute(Event event)
+                 VoteTxt(sent), (int)group->GetLootMethod(), sPlayerbotAIConfig->lootRollLevel, guid.ToString(), itemId);
 		 
-        AnnounceRollChoice(sent, itemId);
+        //AnnounceRollChoice(sent, itemId);
```
Delete comment

kadeshar commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -1224,26 +1188,25 @@ bool MasterLootRollAction::Execute(Event event)
              bot->HasSkill(SKILL_ENCHANTING), IsLikelyDisenchantable(proto));
 		 
-    AnnounceRollChoice(sent, itemId);
+    //AnnounceRollChoice(sent, itemId);
```
Delete comment

Wishmaster117 replied:
done

hermensbas commented:
Can someone else test this ingame?

nugsforthugs commented:
> Can someone else test this ingame? 

I have been testing it since September 2 also. I've been compiling all the latest commits every couple of days, it is working well. No complaints aside from I never see any bots need on anything gear wise.

I am sure if I did raids I would probably see it more there, that's something I still have to test a little bit further but otherwise it works pretty well.

Wishmaster117 commented:
> Can someone else test this ingame?

For those who want to test, i have this branch with lastest playerbots commits and the mart looting system: https://github.com/Wishmaster117/mod-playerbots/tree/Master+ShamanReforge+Smarterloots

kadeshar commented:
I see mulitline statement in for without {}.

Wishmaster117 commented:
> I see mulitline statement in for without {}.

And other problems it's seems

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -112,22 +117,26 @@ static bool IsLockbox(ItemTemplate const* proto)
         for (auto m : mods)
-            if (t == m) return true;
+            if (t == m)
+                return true;
```
solved

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -5,13 +5,1017 @@
+        const bool isProtPal = t.isProtPal || (t.cls == CLASS_PALADIN && t.isTank);  // fallback
+        if (t.isRetPal || isProtPal || t.cls == CLASS_HUNTER || t.isEnhSham)
+        {
+            LOG_DEBUG("playerbots", "[LootPaternDBG] INT+AP group check: prioritaire présent -> {} (spec='{}')",
```
translated

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -5,13 +5,1017 @@
+            return true;  // a priority bot will probably NEED
+        }
+    }
+    LOG_DEBUG("playerbots", "[LootPaternDBG] INT+AP likely-to-need: aucun prioritaire bot en position de NEED");
```
translated

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -22,65 +1026,94 @@ bool LootRollAction::Execute(Event event)
             continue;
 
+        LOG_DEBUG("playerbots",
+                  "[LootRoolDBG] start bot={} item={} \"{}\" class={} q={} lootMethod={} enchSkill={} rp={}",
```
Corrected

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -22,65 +1026,94 @@ bool LootRollAction::Execute(Event event)
 
-        // Armor Tokens are classed as MISC JUNK (Class 15, Subclass 0), luckily no other items I found have class bits and epic quality.
-        if (proto->Class == ITEM_CLASS_MISC && proto->SubClass == ITEM_SUBCLASS_JUNK && proto->Quality == ITEM_QUALITY_EPIC)
+        LOG_DEBUG("playerbots", "[LootRoolDBG] usage={} (EQUIP=1 REPLACE=2 BAD_EQUIP=8 DISENCHANT=13)", (int)usage);
```
Corrected

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -22,65 +1026,94 @@ bool LootRollAction::Execute(Event event)
+            (group->GetLootMethod() == NEED_BEFORE_GREED || group->GetLootMethod() == GROUP_LOOT) &&
+            bot->HasSkill(SKILL_ENCHANTING) && IsLikelyDisenchantable(proto))
+        {
+            LOG_DEBUG("playerbots", "[LootRoolDBG] DE switch: {} -> DISENCHANT (lootMethod={}, enchSkill={}, deOK=1)",
```
Corrected

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -22,65 +1026,94 @@ bool LootRollAction::Execute(Event event)
-            }
+            // Lets CalculateRollVote decide (includes SmartNeedBySpec, BoE/BoU, unique, cross-armor)
+            vote = CalculateRollVote(proto, randomProperty);
+            LOG_DEBUG("playerbots", "[LootRoolDBG] after CalculateRollVote: vote={}", VoteTxt(vote));
```
cprrected

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -5,13 +5,1017 @@
+    return false;
+}
+
+// SPECIAL ITEMS PRIORITY PATERNS //
```
Corrected

Wishmaster117 commented on src/strategy/actions/LootRollAction.cpp:
```diff
@@ -5,13 +5,1017 @@
+    return match;
+}
+
+// Decision: Ret/Hunter/Enh -> primary; non-caster physiques -> not primary, casters -> primary only if no priority in
```
translated

xengine-qyt commented:
this pr is good ,how time to merge

kadeshar commented:
> Not to be nitpicky, but all PRs from here on out also will have to adhere to **Azerothcore's C++ Code Standards** https://www.azerothcore.org/wiki/cpp-code-standards

I dont agree. This PR was made before standarization out. And toghter we agree that PRs before that doesnt need to apply that

Mr-Deadbeat commented:
tested the new feature but there are some strange issues i don't know if it has to do with commands out of sync if there are some overlaps in what the loot system uses compared to the core but selling stuff from chat window "s *" or opening inventory bags trough multibot is not working when i merged to test it out i got no errors during compilation so not sure what is going on atm is the code updated to work with the latest version of mod-playerbots ?

note i'm no coder know some basic stuff but thats about it so maybe i did something wrong during merge of the pull request ?

forgot to add hash for core version c96314a9b3d7f4c4ef0f552ccadb9bf23ba5c027
and used bots 5d3e64800f5281b4cf9d3fc4f25b27bc6a46d1dd at the time of testing

Wishmaster117 commented:
> tested the new feature but there are some strange issues i don't know if it has to do with commands out of sync if there are some overlaps in what the loot system uses compared to the core but selling stuff from chat window "s *" or opening inventory bags trough multibot is not working when i merged to test it out i got no errors during compilation so not sure what is going on atm is the code updated to work with the latest version of mod-playerbots ?
> 
> note i'm no coder know some basic stuff but thats about it so maybe i did something wrong during merge of the pull request ?

Don't noticed that on my tests, I will take a look after my hollidays.

Wishmaster117 commented:
> No you were right. I can't answer your question if only those with the professions are rolling, as all my bots have all professions, and they all roll need for all recipes. I'll keep using this and report back if I see anything odd.

If you can, Try to make a group with 3 addclass-bots .learn a different profession to each one and tell me. When I tested this, only bots who add the profession needed the recipe, if they already have th recipe in bag or if they already know it they GREED.
But as I said I will improve things after this PR merged. I don't want to touch it now because it will be difficult to review. 
For exemple, have to add bots have olso to pass when it's necessary to be more realistic, and also they will need for they second spé if no bot in group/raid potentially will need for first spé, etc...

Mr-Deadbeat commented:
found the problem causing the issues with commands not working i must have had some remnants from old code on the repo i was using deleted everything and cloned fresh both core and playerbots same hash versions and now everything works as expected did notice though that the cross-armor config option does not work at all they are still greeding on stuff even after "disabling" cross-armor option with 9.99 so might want to take a look at that other than that it seems to work as expected

Wishmaster117 commented:
> after further testing cross-armor is a huge problem my shaman is needing on all cloth as enhancement shaman so in essence stealing priest and mage loot
> 
> same goes with shaman and warrior both are stealing rogues leather armor

Don't worry, I'll make them roll as they should. There are exceptions, though; if a paladin bot has nothing in its slot, it's normal for it to roll on cloth gear because it's an upgrade for them. Could you give me more details on exactly what's happening, the bots' specializations, what they're rolling on, etc.?

biship commented:
I have my lvl 30 holy pali running around with cloth and AiPlayerbot.Roll.CrossArmorExtraMargin = 2
So i kind of expected her to be in cloth and haven't dug deeper. I haven't noticed what she is upgrading from however. She had mail at some point, many levels ago.

Mr-Deadbeat commented:
@Wishmaster117 i did a full run of Sunken Temple and there was a lot of rolls to the left and right and so far everything is working way better than before tank got new upgrades so did the other characters as well only thing that bugged me was the priest who has enchanting he wanted to disenchant everything not needed even blue boe items which could benefit others i know one can disable the enchanting roll in the config which i did

also set the default values for the cross-armor setting to 1.20 just enabled the new setting and did not notice anything odd at all will keep testing and see if anything comes up good work thus far :)

Wishmaster117 commented:
> > > @Wishmaster117 i did a full run of Sunken Temple and there was a lot of rolls to the left and right and so far everything is working way better than before tank got new upgrades so did the other characters as well only thing that bugged me was the priest who has enchanting he wanted to disenchant everything not needed even blue boe items which could benefit others i know one can disable the enchanting roll in the config which i did
> > > also set the default values for the cross-armor setting to 1.20 just enabled the new setting and did not notice anything odd at all will keep testing and see if anything comes up good work thus far :)
> > 
> > 
> > Do you mean the priest wanted to disenchant items that were beneficial to him instead of NEED?
> 
> no i don't think he did although i might be wrong did notice today though that my tank took greed instead of need on a plate shoulder piece that had level 45 requirement and tank was level 43 at the time so my guess would be the priest would do the same for items that he can't use but disenchant it instead one of the reasons why i disabled disenchanting just to be safe so BOE or BOU items would be preserved can just store them in bank for future use for whomever it may be ;)
> 
> now when i think about it there was an issue with rogue needing on leather with intellect on it think it had agility or stamina as well
> 
> a question do you check back armor pieces as well the same way you do with other armor or are they excluded from the system ?

Ok have to tweak the desenchant system a little bit and  had a specific condition for rogs to prevent to need on items with Agi and intel.
What do you mean by back armor? Cloak?

Wishmaster117 commented:
Small improvment:
Centralized “who gets priority” into a generic fallback: if no bot in the group has the item as a mainspec upgrade (EQUIP/REPLACE + IsPrimaryForSpec), then offspec upgrades are allowed to keep NEED instead of being forced to GREED.

Healer + Hit items are now handled only via IsPrimaryForSpec (healer treats SP/INT/Hit/no-regen as offspec), and the generic fallback decides whether they can still NEED when no DPS caster bot wants it.

Celandriel commented on conf/playerbots.conf.dist:
```diff
@@ -288,9 +288,62 @@ AiPlayerbot.TwoRoundsGearInit = 0
+# If Need-Before-Greed is active and the item is only useful for disenchanting,
+# bots will press the "Disenchant" button instead of GREED.
+# Default: 1 (enabled)
+AiPlayerbot.Roll.UseDEButton = 1
```
Suggest disabled as default. 
Right now disenchanting items may only serve to clutter bot inventories.

Mr-Deadbeat commented:
using the latest updates and noticed the bots ignore AiPlayerbot.Roll.CrossArmorGreedIsPass=1 completely they need regardless if i forbid it or not atm my hunter just needed on Shadowcraft belt which is leather and he uses mail armor
previously druid needed on cloth armor but that was before i forbid cross armor needing i used extramargin=4.99 at that time to see if something had changed warlock greeded on cloth and druid needed on it thought it might be normal due to it having decent stats and him being restoration spec but this last incident made me question things a bit

Wishmaster117 commented:
@biship , @Mr-Deadbeat can you test this new version?

Wishmaster117 commented:
> > > using the latest updates and noticed the bots ignore AiPlayerbot.Roll.CrossArmorGreedIsPass=1 completely they need regardless if i forbid it or not atm my hunter just needed on Shadowcraft belt which is leather and he uses mail armor previously druid needed on cloth armor but that was before i forbid cross armor needing i used extramargin=4.99 at that time to see if something had changed warlock greeded on cloth and druid needed on it thought it might be normal due to it having decent stats and him being restoration spec but this last incident made me question things a bit
> > 
> > 
> > Okay, I messed up there. I'm making CrossArmorGreedIsPass dependent on a GREED vote and a usage == ITEM_USAGE_BAD_EQUIP, whereas in your example (Shadowcraft Belt for hunter), the actual path goes through: usage = EQUIP, AllowBoENeedIfUpgrade = 1, = direct NEED, without ever going through GREED or BAD_EQUIP. I'll fix that right away.
> 
> just finished a LBRS run and didn't see anything weird going on so it seems to be fixed only thing bugging me is jewelry is not checked against they seem to choose whatever they fancy at the moment lol anywhoo i think it's ready for inclusion as far as i can see but thats up to the powers that be ;) unless @biship has seen something i might have missed

Ok here what's happening:
Jewelry is checked by the new SmartNeedBySpec rules (casters can’t NEED on pure STR rings, physical specs can’t NEED on pure SP necks, tanks get priority on avoidance jewelry, etc.).
What you’re probably seeing is that for very “generic” rings/trinkets (pure DPS stats), several specs are considered main-spec users, so bots are allowed to roll NEED just like real players would.
The cross-armor safety (cloth/leather/mail/plate tier) only applies to body armor by design, so it doesn’t affect jewelry at all.

I added a new helper to handle **jewelry** more precisely, extending the logic *“bot in underwear → others calm down”* to also cover trinkets and cloaks, but only for **NEED off‑spec (fallback)** situations.

I deliberately included cloaks again, but this helper does **not** touch the question of *“is this cloak mainspec or off‑spec?”*  
That part is still handled by `IsPrimaryForSpec` and `IsFallbackNeedReasonableForSpec`.

What the helper adds is a **group rule**:  
If another bot in the group has the jewelry/cloak slot empty or only grey/white gear and sees the item as an upgrade, then the off‑spec fallback NEED of the others on that piece is downgraded → they roll GREED instead.

In other words:  
Previously, for a cloak like **43305**, the logic was only:  
*“is this hunter / ret paladin / prot paladin allowed to NEED, or is it an unreasonable off‑spec?”*  

Now, in addition, we ask:  
*“OK, this hunter could NEED in fallback… but is someone else in the group naked in this slot? If yes → we give them moral priority, and the other off‑specs drop to GREED.”*

This logic did not exist for cloaks before:  
`GroupHasDesperateUpgradeUser` (cross‑armor version) only applies to body armor (HEAD, CHEST, LEGS, etc.), and `IsBodyArmorInvType` explicitly excludes cloaks (no `INVTYPE_CLOAK` inside).

The whole *“stat profile / mainspec vs off‑spec”* logic already covers cloaks via `isJewelry`.  
The new helper does not duplicate that; it solves a different problem: *“who is naked in this slot within the group?”*

Cloaks were not covered by the *“desperate upgrade user”* logic before. 

So @biship and @Mr-Deadbeat can you test this now please?

Wishmaster117 commented:
@Mr-Deadbeat , @biship Thank you for your tests with your feedback, it allows me to make the bots’ roll as accurate as possible.
I made some changes:
This changes tightens the loot roll AI so that bots respect armor proficiencies and make smarter decisions around Spirit-based jewelry. The goal is to avoid obviously wrong NEED rolls (e.g. a mage needing on leather shoulders, or a warlock sniping a pure Spirit healer necklace) while keeping the existing SmartNeedBySpec and StatsWeightCalculator framework intact.

**Armor proficiency & cross-armor behavior**
Previously, the loot roll logic only looked at class/role and stat profiles to decide whether an item was “primary spec”, with a one-way cross-armor guard that prevented plate/mail users from NEED-ing on cloth/leather. Cloth wearers, however, were allowed to NEED on leather/mail/plate as long as the stats looked “caster-like”.

This patch introduces an explicit armor proficiency check:

- For any body armor piece (head/shoulders/chest/legs/feet/wrists/hands/waist), we now verify that the bot actually has the corresponding armor skill:

- Leather → SKILL_LEATHER
- Mail → SKILL_MAIL
- Plate → SKILL_PLATE_MAIL

- If the bot does not have the skill, that item is never treated as a primary spec upgrade, regardless of stats.
- The same rule is applied in the fallback logic: if the bot cannot even wear the armor, fallback NEED is considered never reasonable.

Concretely:

- Cloth casters (mage/warlock/priest) will no longer NEED mainspec on leather/mail/plate pieces, even if they carry perfect caster stats.
- Low-level warriors/paladins/hunters/shamans will not NEED plate/mail before they actually learn the proficiency.
- Cross-armor NEED remains possible only via the existing cross-armor upgrade heuristic, and only when:

- the item is lower-tier armor,
- no proper primary armor user in the group is likely to need it, and
- the upgrade is large enough according to StatsWeightCalculator.

This makes NEED strictly aligned with what the character can wear and is supposed to wear.

**Spirit jewelry: healer-oriented, not DPS candy**

The second part of the change refines how we classify Spirit-heavy jewelry (rings/necklaces/trinkets/cloaks):

- I introduce a clear profile for “pure Spirit regen jewelry”:

- isJewelry == true
- has Spirit
- no Intellect
- no Spell Power
- no offensive ratings (Hit/Haste/Crit)
- no physical stats (STR/AGI/AP/ARP/EXP)

This profile is treated as:

- Healer mainspec only:

- Healer specs (holy paladin, resto druid/shaman, non-shadow priests) may consider these items mainspec.

- Off-spec for DPS casters:

- DPS casters (mage, warlock, shadow priest, etc.) will not treat pure Spirit jewelry as mainspec.
- They are also explicitly blocked from using fallback NEED on these pieces: even if no healer wants it as mainspec, DPS casters are limited to GREED.

In practice:

- A warlock or mage will no longer NEED on a “Spirit-only” necklace over a proper healer, even if the raw scoring sees it as a theoretical upgrade.
- Healers keep priority on Spirit regen jewelry, matching how players usually distribute these items in groups/raids.

**Fallback NEED rules made stricter**

The generic “fallback NEED” rules, used when an item is not mainspec but there is no clear primary-spec candidate in the group, are now more conservative:

- If the bot cannot wear the armor type (missing proficiency skill), fallback NEED is rejected outright.
- DPS casters are not allowed to fallback-NEED on pure Spirit regen jewelry (reserved for healers).
- Existing protections remain:

- Physical specs do not fallback-NEED pure caster armor/SP weapons.
- Casters/healers do not fallback-NEED pure physical armor or melee-only jewelry.

This keeps the “off-spec NEED if nobody really needs it” behavior, but closes the most egregious loopholes (like a mage needing leather shoulders, or a warlock taking healer jewelry “because it’s technically an upgrade”).

**Behavior impact**

With these changes:

- Bots will no longer NEED on armor types they are not supposed to wear.
- Cloth casters stop NEED-ing on leather/mail/plate caster pieces as mainspec.
- Spirit-only jewelry is prioritized for healers; DPS casters will GREED it instead of NEED-ing.
- Cross-armor NEED remains available only in very limited, strongly justified upgrade scenarios.

Overall, loot rolls should now align much better with human expectations: mainspec first, correct armor types only, and healer-oriented regen gear reserved for healing specs.
